File: LC/[0376] Wiggle Subsequence - Greedy II.cpp
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        int ret=1;
        int dir=-2; //special assignment for initial slope

        if (nums.size()==0) return 0;
        for(int ii=1; ii < nums.size(); ++ii){
            int dir_prev = dir;
            if(nums[ii] > nums[ii-1])//find out current slope
                dir=1;
            else if (nums[ii] < nums[ii-1])
                dir=-1;
            else //inheriting prev non-zero slope
                dir=dir_prev;
            
            if(dir != dir_prev)
                ret += 1;
        }
        return ret;
    }
};

File: LC/[0376] Wiggle Subsequence - Greedy II.py
class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        N=len(nums)
        dir=-2
        cnt=1
        for ii in range(1, N):
            prev_dir=dir
            if nums[ii] > nums[ii-1]:
                dir=1
            elif nums[ii] < nums[ii-1]:
                dir=-1 

            if dir != prev_dir:
                cnt+=1        

        return cnt